agent/allow_list.go

```diff
@@ -0,0 +1,81 @@
+package agent
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+type allowList struct {
+	Exec      string   `yaml:"exec" json:"exec"`
+	Args      []string `yaml:"args" json:"args"`
+	Sha512    string   `yaml:"sha512" json:"sha512"`
+	EnableEnv bool     `yaml:"enable_env" json:"enable_env"`
+}
+
+func readAllowList(path string, readBytes func(string) ([]byte, error)) ([]allowList, error) {
+	var allowList []allowList
+	if path == "" {
+		return allowList, nil
+	}
+	unmarshalFuncs := map[string]func(in []byte, out interface{}) error{
+		".yaml": yaml.Unmarshal,
+		".yml":  yaml.Unmarshal,
+		".json": json.Unmarshal,
+	}
+	for ext, f := range unmarshalFuncs {
+		if strings.Contains(path, ext) {
+			bytes, err := readBytes(path)
+			if err != nil {
+				return nil, err
+			}
+			err = f(bytes, &allowList)
+			if err != nil {
+				return nil, err
+			}
+			for _, al := range allowList {
+				err = al.validate()
+				if err != nil {
+					return nil, err
+				}
+			}
+			return allowList, nil
+		}
+	}
+
+	return nil, fmt.Errorf("invalid file extension")
+}
+
+// validate returns an error if the allowList contains invalid values.
+func (al *allowList) validate() error {
+	if al.Exec == "" {
+		return errors.New("exec cannot be empty")
+	}
+
+	if len(al.Args) == 0 {
+		return errors.New("args cannot be empty")
+	}
+
+	return nil
+}
+
+func (a *Agent) matchAllowList(command string) (allowList, bool) {
+	for _, al := range a.allowList {
+		remaining := command
+		if strings.Contains(command, al.Exec) {
+			remaining = strings.Trim(remaining, al.Exec)
```

Comment:
This is probably not what you want, and could lead to incorrect results. Unlike TrimPrefix and TrimSuffix, Trim uses a set of runes to be trimmed, not a whole string.

See the example here to see what I'm talking about: https://golang.org/pkg/strings/#Trim

Reply:
The intention is to trim leading and trailing whitespace.

Reply:
Trim doesn't trim whitespace, that's https://golang.org/pkg/strings/#TrimSpace

Reply:
Oops I misread this. The intention is to remove the substring `al.Exec` from the string `remaining`.

Reply:
After reading through the example I see what you mean!

Reply:
I mean, obviously Trim will trim whitespace if you give it whitespace to trim, but it will trim the entire string of whitespace, not only trailing and leading :)